### ISSUE TITLE
daemon startup should not access /root/.local

### ISSUE THREAD

**Issue by sylvestre** [fail2ban/fail2ban#2688]
Reported here: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=956177

Seems that fail2ban is trying to access /root/.local/
```
type=AVC msg=audit(1586313861.749:37): avc:  denied  { search } for  pid=704 comm="fail2ban-server" name=".local" dev="sdb2" ino=31516 scontext=system_u:system_r:fail2ban_t:s0 tcontext=unconfined_u:object_r:xdg_data_t:s0 tclass=dir permissive=0
```

Maybe fail2ban should add an option to disable this behavior (or not to do it by default)

**Comment by sebres** [comment]
Hmm... this is strange.
Fail2ban would not try to access such file directly. The only one thing I can imagine where it could try to access `.local` file would be the a dynamically constructed config-files with empty base name (where normally it is rather switching from `*.conf` to `*.local` and the base-name is never empty, or it is related to `/etc/fail2ban`). 
Anyway neither I see where exactly some include or config file would produce such file name (also if it would be relative current directory and it is `/root/` for some reasons), nor I can reproduce it trying current 0.10 / 0.11 versions.

I think we'd need more info here:
1. I guess fail2ban is started by sustemd-unit, please provide the exact command-line of that;
2. please provide the modifications made in fail2ban configuration as well as the output of:
`fail2ban-client -vvvd | grep .local`
3. is it also reproducible if fail2ban will be started directly (e. g. by `?sudo? /fail2ban-server -f start`) from some directory (like `/tmp`) and from `/root/` as current directory)?

**Comment by sylvestre** [comment]
@sebres I asked the reporter of the Debian bug to reply here, thanks!

**Comment by etbe** [comment]
stat("/root/.local/lib/python3.8/site-packages", 0x7ffe878316d0) = -1 ENOENT (No such file or directory)

Above is the access in question.

/usr/bin/python3 /usr/bin/fail2ban-server -xf start

Above is the start command. It happens if I run that from the command-line with /tmp or with /root as the current directory.

**Comment by sebres** [comment]
Oh... this is the local [site-packages "feature"](https://stackoverflow.com/questions/31384639/what-is-pythons-site-packages-directory)... For some reason you'd have home/.local in path or PYTHONPATH (or PYTHONNOUSERSITE) environment variable set.
What do you see here:
```bash
?sudo? fail2ban-python -c 'import sys; print(sys.path)'
?sudo? fail2ban-python -m site --user-site
```
As for sys.path, for debian's it should normally not contain site-packages at all (they using `dist-packages` to install their distributed modules). But maybe it was changed in some python version (3.8?)..
Other distributions using default standard location for the site-packages as `prefix/lib/pythonX.Y/...`, where prefix is mostly `/usr/local/lib` or `/usr/lib`.

As for `site` module, neither fail2ban uses `site` module directly, nor it is a requirement of some modules that fail2ban using at the moment.

But since python has 2 parameters to disable site directory, this could mean such inclusion may happen per default:
  - `-s`     : don't add user site directory to sys.path; also PYTHONNOUSERSITE
  - `-S`     : don't imply 'import site' on initialization

To test this, can you please try to start fail2ban-server (or fail2ban-client) as following?
```
?sudo? fail2ban-python -sS $(whereis fail2ban-server) -f start
```
Would that solve the issue?

**Comment by sebres** [comment]
There is also `PYTHONNOUSERSITE` environment variable, from the documentation:
> If this is set, Python won’t add the user site-packages directory to sys.path.

So you could also try to add `Environment="PYTHONNOUSERSITE=yes"` to Service section of your fail2ban systemd unit or even in override file, for instance `systemctl edit fail2ban` and add this:
```ini
[Service]
Environment="PYTHONNOUSERSITE=yes"
```

**Comment by etbe** [comment]
On Wednesday, 15 April 2020 7:59:45 PM AEST Sergey G. Brester wrote:
> Oh... this is the local [site-packages
> "feature"](https://stackoverflow.com/questions/31384639/what-is-pythons-sit
> e-packages-directory)... For some reason you'd have home/.local in path or
> PYTHONPATH (or PYTHONNOUSERSITE) environment variable set. What do you see
> here:
> ```bash
> ?sudo? fail2ban-python -c 'import sys; print(sys.path)'

# fail2ban-python -c 'import sys; print(sys.path)'
['', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-
dynload', '/usr/local/lib/python3.8/dist-packages', '/usr/lib/python3/dist-
packages']

> ?sudo? fail2ban-python -m site --user-site

# fail2ban-python -m site --user-site
/root/.local/lib/python3.8/site-packages

> As for sys.path, for debian's it should normally not contain site-packages
> at all (they using `dist-packages` to install their distributed modules).
> But maybe it was changed in some python version (3.8?).. Other
> distributions using default standard location for the site-packages as
> `prefix/lib/pythonX.Y/...`, where prefix is mostly `/usr/local/lib` or
> `/usr/lib`.
> 
> As for `site` module, neither fail2ban uses `site` module directly, nor it
> is a requirement of some modules that fail2ban using at the moment.
> 
> But since python has 2 parameters to disable site directory, this could mean
> such inclusion may happen per default: - `-s`     : don't add user site
> directory to sys.path; also PYTHONNOUSERSITE - `-S`     : don't imply
> 'import site' on initialization
> 
> To test this, can you please try to start fail2ban-server (or
> fail2ban-client) as following? ```
> ?sudo? fail2ban-python -sS $(whereis fail2ban-server) -f start
> ```
> Would that solve the issue?

# fail2ban-python -sS /usr/bin/fail2ban-server -xf start
Traceback (most recent call last):
  File "/usr/bin/fail2ban-server", line 34, in <module>
    from fail2ban.client.fail2banserver import exec_command_line, sys
ModuleNotFoundError: No module named 'fail2ban'

That didn't work.

But "fail2ban-python -s /usr/bin/fail2ban-server -xf start" worked and made it 
stop accessing /root/.local.  So it seems that -s solved the problem and -S 
caused some other problem.


-- 
My Main Blog         http://etbe.coker.com.au/
My Documents Blog    http://doc.coker.com.au/

**Comment by sebres** [comment]
> So it seems that -s solved the problem and -S caused some other problem.

Can you try that with env-variable as suggested in https://github.com/fail2ban/fail2ban/issues/2688#issuecomment-613949188 ?

**Comment by etbe** [comment]
On Wednesday, 15 April 2020 8:13:19 PM AEST Sergey G. Brester wrote:
> There is also `PYTHONNOUSERSITE` environment variable, from the
> documentation:
> > If this is set, Python won’t add the user site-packages directory to
> > sys.path.
> 
> So you could also try to add `Environment="PYTHONNOUSERSITE=yes"` to Service
> section of your fail2ban systemd unit or even in override file, for
> instance `systemctl edit fail2ban` and add this:
 ```ini
> [Service]
> Environment="PYTHONNOUSERSITE=yes"

Yes that works too.

-- 
My Main Blog         http://etbe.coker.com.au/
My Documents Blog    http://doc.coker.com.au/

**Comment by sebres** [comment]
Then fixed in 7e3061e7ace0e973378a17de04a2692142e1a6c1 for systemd unit file (other init scripts can be updated on demand, because it is rather the matter of maintainers).
